lib/matplotlib/backend_bases.py

```diff
@@ -142,6 +149,35 @@ def get_registered_canvas_class(format):
     return backend_class
 
 
+class MainLoopBase(object):
+    """This gets used as a key maintaining the event loop.
+    Backends should only need to override begin and end.
+    It should not matter if this gets used as a singleton or not due to
+    clever magic.
+    """
+    _instance_count = {}
+    _running = False
+    def __init__(self):
+        MainLoopBase._instance_count.setdefault(self.__class__, 0)
+        MainLoopBase._instance_count[self.__class__] += 1
+
+    def begin(self):
```

Comment:
As mentioned in the other PR this should probably be a contextmanager instead, which also avoids the need of "clever magic".